src/java/com/zimbra/clam/ClamScanner.java

```diff
@@ -102,73 +129,85 @@ protected Result accept(InputStream is, StringBuffer info) {
+            outStream.flush();
 
-            if (LOG.isDebugEnabled()) { LOG.debug("stream connect to " + mClamdHost + ":" + port); }
-            dataSocket = new Socket(mClamdHost, port);
-            if (data != null) {
-                dataSocket.getOutputStream().write(data);
-                if (LOG.isDebugEnabled()) { LOG.debug("wrote " + data.length + " bytes"); }
-            } else {
-                long count = ByteUtil.copy(is, false, dataSocket.getOutputStream(), false);
-                if (LOG.isDebugEnabled()) { LOG.debug("copied " + count + " bytes"); }
+            // if byte array passed as input instead of inputStream
+            if (data != null && inputStream == null) {
+                inputStream = new ByteArrayInputStream(data);
             }
-            dataSocket.close();
 
-            if (LOG.isDebugEnabled()) { LOG.debug("reading result"); }
-            String answer = in.readLine();
-            if (answer == null) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("sending data for scanning, chunkSize: " + chunkSize);
+            }
+            byte[] buffer = new byte[chunkSize];
+            socketInputStream = socket.getInputStream();
+            int read = inputStream.read(buffer);
+            while (read >= 0) {
+                byte[] chunkSize = ByteBuffer.allocate(4).putInt(read).array();
+                outStream.write(chunkSize);
+                outStream.write(buffer, 0, read);
+                if (socketInputStream.available() > 0) {
+                    byte[] reply = readAll(socketInputStream);
+                    throw new IOException("Reply from server: " + new String(reply, StandardCharsets.ISO_8859_1));
+                }
+                read = inputStream.read(buffer);
+            }
+            outStream.write(new byte[]{0, 0, 0, 0});
+            outStream.flush();
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("reading result");
+            }
+            String answer = new String(readAll(socketInputStream), StandardCharsets.ISO_8859_1);
+            LOG.debug("answer =" + answer);
+            if (StringUtil.isNullOrEmpty(answer)) {
                 throw new ProtocolException("EOF from clamd when looking for result");
             }
             info.setLength(0);
-            if (answer.startsWith("stream: ")) {
-                answer = answer.substring("stream: ".length());
+            String extractedAns = "";
+            if (answer.startsWith(RESULT_PREFIX)) {
+                extractedAns = answer.substring(RESULT_PREFIX.length()).trim();
+                LOG.debug("extracted answer: " + extractedAns);
             }
-            info.append(answer);
-            if (answer.equals("OK")) {
+            info.append(extractedAns);
+            if (ANSWER_OK.equals(extractedAns)) {
                 return ACCEPT;
             } else {
                 return REJECT;
```

Comment:
we should add some logging here so that we know which user tried which affected file attachment, so we have sufficient logging

Reply:
Exception gets thrown in callee when it founds attachment gets rejected, you will see a long stack trace there with the file name and ClamAV response.